sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala

```diff
@@ -299,13 +401,16 @@ class PrefixKeyScanStateEncoder(
  * @param keySchema - schema of the key to be encoded
  * @param orderingOrdinals - the ordinals for which the range scan is constructed
  * @param useColumnFamilies - if column family is enabled for this encoder
+ * @param avroEnc - if Avro encoding is specified for this StateEncoder, this encoder will
+ *                be defined
  */
 class RangeKeyScanStateEncoder(
     keySchema: StructType,
     orderingOrdinals: Seq[Int],
     useColumnFamilies: Boolean = false,
-    virtualColFamilyId: Option[Short] = None)
-  extends RocksDBKeyStateEncoderBase(useColumnFamilies, virtualColFamilyId) {
+    virtualColFamilyId: Option[Short] = None,
+    avroEnc: Option[AvroEncoder] = None)
```

Comment:
Instead of avroEnc, I would honestly introduce another interface:

```scala
trait Serde {

  def encodeToBytes(...)

  def decodeToUnsafeRow(...)
  
  def encodePrefixKeyForRangeScan(...)

  def decodePrefixKeyForRangeScan(...)
}
```
and move the logic in there so that you don't have to keep on doing `avroEnc.isDefined` for these

The logic seems pretty similar except for the input data. The AvroStateSerde or whatever you want to name it would have the `private lazy val remainingKeyAvroType = SchemaConverters.toAvroType(remainingKeySchema)`

Reply:
Spoke offline - it doesn't look like this simplifies things an awful lot - can be a follow-up.